Support pseudo selectors on `elements` in theme json

getdave opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
As part of work towards https://github.com/WordPress/gutenberg/issues/38277, this PR enables `theme.json` support for `elements` (both top and _block_ level) to contain **pseudo selectors** in the form:

```json
{
    "elements": {
        "link": {
            "color": {
                "text": "green"
            },
            ":hover": {
                "color": {
                    "text": "hotpink"
                }
            }
        }
    },
    "blocks": {
        "core/group": {
             "elements": {
                "link": {
                    ":hover": {
                        "color": {
                            "text": "red"
                        }
                    }
                }
            }
        }
    }
}
```

Currently this PR imposes a whitelist which governs which psuedo selectors are available to be used on which elements. It looks like this but could be augmented as required:

```php
const VALID_ELEMENT_PSEUDO_SELECTORS = array(
    'link' => array( ':hover', ':focus' ),
);
```

Please note that once an element is "open" to a given pseudo selector all valid stlye properties are available to be changed. For example the following is valid whether you _should_ do it or not:

```json
{
    "elements": {
        "link": {
            "color": {
                "text": "green"
            },
            ":hover": {
                "color": {
                    "text": "hotpink"
                },
                "typography": {
                    "textTransform": "uppercase",
                    "fontSize": "10em"
                }
            }
        }
    }
}
```

My personal preference would be to lock things down to a bear minimum to avoid a11y issues but I guess perhaps we'll want things open to developers to make those choices just like they do every day in CSS. We can limit user choices but choosing not to expose UI controls for every possible style property.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

This is low level work to support styling of interactivity states in the editor and front end.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Update the sanitization routines to allow various pseudo selectors on a per `element` basis. For now we only support `link`.

## Todo

- [x] Get `WP_Theme_JSON_Gutenberg::remove_insecure_properties` to allow for pseudo selectors.
- [x] Ensure pseudo selectors end up in final processed theme json
- [x] Ensure correct styles are generated for pseudo selectors from the theme json in editor.
- [x] Ensure correct styles are generated for pseudo selectors from the theme json on front end.
- [x] Consider enabling block-level `elements` pseudo selector support.
- [ ] Learn how to type/spell `pseudo` (note: this task will never be completed)
- [x] Add/augment automated tests.
- [ ] ~Create an allow/whitelist for per element pseudo selectors for use on client side (or pass the existing config value from PHP)~ - this can be handled in a followup.
- [x] Augment frontend JS global styles tests to assert on pseudo selector generation.
- [x] Refine PR desc and make RfR

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

### Stress testing

Note this PR is being stress tested by [applying it various Themes](https://github.com/Automattic/themes/issues/6108).

### Testing with Archeo test PR

I have created an [example PR on the Archeo theme](https://github.com/Automattic/themes/pull/6101) to test this out. You can try that by running a copy of Archeo with that PR checked out.

### Manual testing
- Use a Theme where you can modify the `theme.json` file.
- Try adding some psudo selector styles for `link` element at top level:
```json
{
    "elements": {
        "link": {
            "color": {
                "text": "green"
            },
            ":hover": {
                "color": {
                    "text": "hotpink"
                }
            }
        }
    }
}
```
- validate standard `a` elements have a `hotpink` color when you hover over them on _both_ editor and frontend.
- now add an _invalid_ pseudo selector and check it isn't output in the stylesheet
```json
{
    "elements": {
        "link": {
            "color": {
                "text": "green"
            },
            ":thisisaninvalidselector": {
                "color": {
                    "text": "hotpink"
                }
            }
        }
    }
}
```
- following the same pattern as above now try adding an _unsupported_ selector such as `:target`.
- following the same pattern as above now try adding an _unsupported_ element like `button`. 
```json
{
    "elements": {
        "button": {
            "color": {
                "text": "green"
            },
            ":hover": {
                "color": {
                    "text": "hotpink"
                }
            }
        }
    }
}
```

Please be sure to check in both the Post and the Site Editor.

## Screenshots or screencast <!-- if applicable -->

### Editor
![Screen Capture on 2022-06-17 at 16-00-13](https://user-images.githubusercontent.com/444434/174335218-7a721265-f211-4426-846a-5c0a348668f3.gif)

### Frontend
![Screen Capture on 2022-06-17 at 15-57-50](https://user-images.githubusercontent.com/444434/174335230-5fc82522-b8a3-47fd-a414-eab50077a0f0.gif)


## Authors



Co-authored-by: Andrei Draganescu <[me@andreidraganescu.info](mailto:me@andreidraganescu.info)>, Adam Zielinski, Ben Dwyer <[ben@scruffian.com](mailto:ben@scruffian.com)>

getdave commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -27,6 +31,205 @@ class WP_Theme_JSON_6_1 extends WP_Theme_JSON_6_0 {
+		$schema_styles_elements         = array();
+		$schema_styles_pseudo_selectors = array();
+
+		// Set allowed element pseudo selectors based on per element allow list.
+		foreach ( $valid_element_names as $element ) {
+			$schema_styles_elements[ $element ] = $styles_non_top_level;
+
+			if ( array_key_exists( $element, static::VALID_ELEMENT_PSEUDO_SELECTORS ) ) {
+				foreach ( static::VALID_ELEMENT_PSEUDO_SELECTORS[ $element ] as $selector ) {
+					$schema_styles_elements[ $element ][ $selector ] = $styles_non_top_level;
+				}
+			}
+		}
```
Here we add a new sub item for each pseudo selector to the schema for each `element` (e.g. `link`, `h1`...etc). Each selector is the key with the valid style values as the value.

getdave commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -27,6 +31,205 @@ class WP_Theme_JSON_6_1 extends WP_Theme_JSON_6_0 {
+		$schema['settings']           = static::VALID_SETTINGS;
+		$schema['settings']['blocks'] = $schema_settings_blocks;
+
+		// TARGET STRUCTURE
+		// $schema['styles']['elements']['link'][':hover'] = // allowed pseudo selectors FTW
+		// $schema['styles']['blocks']['core/button']['elements']['link'][':hover'] = // allowed pseudo selectors for block sub elements FTW
```
We should remove this or add as a comment above the relevant places.

getdave commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -27,6 +31,205 @@ class WP_Theme_JSON_6_1 extends WP_Theme_JSON_6_0 {
+		foreach ( $valid_block_names as $block ) {
+			$schema_settings_blocks[ $block ]           = static::VALID_SETTINGS;
+			$schema_styles_blocks[ $block ]             = $styles_non_top_level;
+			$schema_styles_blocks[ $block ]['elements'] = $schema_styles_elements;
```
We'll need to update the block elements schema to allow for pseudo selectors as well. In this PR?

getdave replied:
✅

getdave commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -27,6 +31,205 @@ class WP_Theme_JSON_6_1 extends WP_Theme_JSON_6_0 {
+				continue;
+			}
+
+			$result = static::remove_keys_not_in_schema( $input[ $subtree ], $schema[ $subtree ] );
```
This line is _why_ we need to update the schema to allow for pseudo selectors.

getdave commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -27,6 +31,205 @@ class WP_Theme_JSON_6_1 extends WP_Theme_JSON_6_0 {
+	 * @param array $input Node to process.
+	 * @return array
+	 */
+	protected static function remove_insecure_styles( $input ) {
```
I think we can safely remove this function and leave it to the version in WP Core. We've not modified it right?

ramonjd replied:
If it's not been modified between 5.9 and 6.0 I _think_ so, though I'd hate to be quoted on that 😄 

Should we wait for this PR first? https://github.com/WordPress/gutenberg/pull/41306

Is there a reason why it's in 6.1 here if it's not been altered?

getdave commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -27,6 +31,205 @@ class WP_Theme_JSON_6_1 extends WP_Theme_JSON_6_0 {
+
+			$output = static::remove_insecure_styles( $input );
+
+			// Get a reference to element name from path.
+			// $metadata['path'] = array('styles','elements','link');
+			$current_element = $metadata['path'][ count( $metadata['path'] ) - 1 ];
+
+			// $output is stripped of pseudo selectors. Readd and process them
+			// for insecure styles here.
+			if ( isset( static::VALID_ELEMENT_PSEUDO_SELECTORS[ $current_element ] ) ) {
+
+				foreach ( static::VALID_ELEMENT_PSEUDO_SELECTORS[ $current_element ] as $pseudo_selector ) {
+					if ( isset( $input[ $pseudo_selector ] ) ) {
+						$output[ $pseudo_selector ] = static::remove_insecure_styles( $input[ $pseudo_selector ] );
+					}
+				}
+			}
```
Added - here we check whether the current element is in the allowed list for pseudo selectors.

If it is then we loop over all the whitelisted selectors for the given element and append them to the output (having first removed insecure styles as for "normal" properties).

This allows for the following structure to be returned from `remove_insecure_properties`:

```php
array(
    'styles'  => array(
        'elements' => array(
            'link' => array(
                'color'  => array(
                    'text'       => 'hotpink',
                    'background' => 'yellow',
                ),
                ':hover' => array(
                    'color' => array(
                        'text'       => 'red',
                        'background' => 'blue',
                    ),
                ),
            ),
        ),
    ),
);
```

Without this change the `:hover` key would not be preserved.

getdave commented:
## Next Steps

It appears that each psuedo selector is duplicated in the final CSS stylesheet. For example:

```css
a {
    background-color: red;
    color: green;
}

a:hover {
    background-color: red;
    color: green;
}

a:hover {
    background-color: green;
    color: red;
}

a:focus {
    background-color: red;
    color: green;
}

a:focus {
    background-color: black;
    color: yellow;
}
```

getdave commented:
## Next Steps

Currently the pseudo selector rules are not being output in the Site Editor. It all works in the Post Editor and also o the front end.

Chief suspect `gutenberg_get_global_stylesheet` as this is running on the Site Editor page and also calls `get_stylesheet()` from the Theme JSON main class.

Ends up in 

https://github.com/WordPress/wordpress-develop/blob/ad4a3ad33eb251c174917aebc0b4637c9565bc9e/src/wp-admin/site-editor.php#L105-L113

ramonjd commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -27,6 +31,244 @@ class WP_Theme_JSON_6_1 extends WP_Theme_JSON_6_0 {
+		// Build the schema based on valid block & element names.
+		$schema                         = array();
+		$schema_styles_elements         = array();
+		$schema_styles_pseudo_selectors = array();
```
Is this needed? Wasn't sure. I know it's early days... :)

getdave replied:
Probably not. We were adding lots of stuff for debugging that probably now can be tidied away.

getdave replied:
✅ Removed. Thanks for the spot.

ramonjd commented:
> Learn how to type/spell pseudo

When you do, can you teach me? 😄 

> Currently the pseudo selector rules are not being output in the Site Editor. It all works in the Post Editor and also o the front end.

This might be a long shot, but I was also having trouble with overwriting core's `get_block_editor_settings()` in Gutenberg

https://github.com/WordPress/gutenberg/pull/39529#issuecomment-1154709502

ramonjd commented:
> Please note that once an element is "open" to a given pseudo selector all valid stlye properties are available to be changed.

If this is limited to theme.json, and not available to controls via the UI as you mention, then I wouldn't have a problem with it as it gives more control to theme developers.

> I guess perhaps we'll want things open to developers to make those choices just like they do every day in CSS. We can limit user choices but choosing not to expose UI controls for every possible style property.

If the intention is to **only** add color UI controls for `link:hover` then we have a relationship between the theme.json API and the editor: users can override link hover colors defined theme.json via global or block styles.

Would there be any benefit over regular CSS to be able to add extra styles such as `typography` or whatever within a `:hover` block in theme.json?

I suppose I'm also wondering if we should restrict/ignore styles with the exception of `color` within `elements.link.[':hover']`

getdave commented:
## Status

This PR is now working in the sense that styles are outputted based on Theme JSON in all editors and on the front end.

Refine `toStyles` from `packages/edit-site/src/components/global-styles/use-global-styles-output.js` so that it dedupes default selectors and pseudo selectors when generating rulesets.

We should be able to mirror the way this is done in PHP by checking if the given selector is a pseudo selector and generating rules accordingly rather than having to have a totally separate conditional branch.

getdave commented:
Thanks for your 👀 on this @ramonjd 🙇‍♂️ 

> If this is limited to theme.json, and not available to controls via the UI as you mention, then I wouldn't have a problem with it as it gives more control to theme developers.

Agreed - it would be limited to Theme JSON. We will only expose UI for very specific elements and pseudo selectors.

> If the intention is to **only** add color UI controls for `link:hover` then we have a relationship between the theme.json API and the editor: users can override link hover colors defined theme.json via global or block styles.

As I envisage it this is what I'm describing above when I say Theme JSON will be open but the UI will be limited. Did I understand you correctly?

> Would there be any benefit over regular CSS to be able to add extra styles such as `typography` or whatever within a `:hover` block in theme.json?

I'm not sure I see one, other than that it would allow all styling related info in a single place instead of split between `theme.json` for some things and `style.css` for others.

If we want to restrict this we're actually going to have to write _more_ code than if we don't want to. I'd be interested in wider opinions on this though.

> I suppose I'm also wondering if we should restrict/ignore styles with the exception of `color` within `elements.link.[':hover']`

Yes this is a good question. This is for individual block styles right? These are only configurable via the UI correct? In which case we'd want to lock these down to a limited subset of values which the project deems are suitable. The criteria for this would be based on how it impacts a11y, user demand .etc. Again please correct me if I have the wrong end of the stick.

> When you do, can you teach me? 😄

I'd have to charge 😜

getdave commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -222,11 +430,32 @@ private static function get_block_nodes( $theme_json, $selectors = array() ) {
+		$settings = _wp_array_get( $this->theme_json, array( 'settings' ) );
+
+		// Attempt to parse a pseudo selector (e.g. ":hover") from the $selector ("a:hover").
+		$pseudo_matches = array();
+		preg_match( '/:[a-z]+/', $selector, $pseudo_matches );
+		$pseudo_selector = $pseudo_matches[0] ?? null;
```
👀 needed here on the regex which intends to match a pseudo selector within a selector.

adamziel replied:
This won't correctly match pseudoelements like `::first-letter` and mixed states such as `:active:hover`, but it also doesn't have to at this point. Other than that, it looks good.

getdave commented on packages/edit-site/src/components/global-styles/test/use-global-styles-output.js:
```diff
@@ -358,27 +439,12 @@ describe( 'global styles renderer', () => {
 				},
 				'core/heading': {
 					selector: 'h1,h2,h3,h4,h5,h6',
-					elements: {
-						link:
-							'h1 ' +
-							ELEMENTS.link +
-							',h2 ' +
-							ELEMENTS.link +
-							',h3 ' +
-							ELEMENTS.link +
-							',h4 ' +
-							ELEMENTS.link +
-							',h5 ' +
-							ELEMENTS.link +
-							',h6 ' +
-							ELEMENTS.link,
-					},
```
I could see no reason that we needed this. The element selector rules seemed to be generated ok without it.

getdave commented on packages/edit-site/src/components/global-styles/test/use-global-styles-output.js:
```diff
@@ -358,27 +439,12 @@ describe( 'global styles renderer', () => {
 				},
 				'core/heading': {
 					selector: 'h1,h2,h3,h4,h5,h6',
-					elements: {
-						link:
-							'h1 ' +
-							ELEMENTS.link +
-							',h2 ' +
-							ELEMENTS.link +
-							',h3 ' +
-							ELEMENTS.link +
-							',h4 ' +
-							ELEMENTS.link +
-							',h5 ' +
-							ELEMENTS.link +
-							',h6 ' +
-							ELEMENTS.link,
-					},
 				},
 			};
 
 			expect( toStyles( tree, blockSelectors ) ).toEqual(
 				'body {margin: 0;}' +
-					'body{background-color: red;margin: 10px;padding: 10px;}h1{font-size: 42px;}.wp-block-group{margin-top: 10px;margin-right: 20px;margin-bottom: 30px;margin-left: 40px;padding-top: 11px;padding-right: 22px;padding-bottom: 33px;padding-left: 44px;}h1,h2,h3,h4,h5,h6{color: orange;}h1 a,h2 a,h3 a,h4 a,h5 a,h6 a{color: hotpink;}' +
+					'body{background-color: red;margin: 10px;padding: 10px;}h1{font-size: 42px;}a{color: blue;}a:hover{color: orange;}a:focus{color: orange;}.wp-block-group{margin-top: 10px;margin-right: 20px;margin-bottom: 30px;margin-left: 40px;padding-top: 11px;padding-right: 22px;padding-bottom: 33px;padding-left: 44px;}h1,h2,h3,h4,h5,h6{color: orange;}h1 a,h2 a,h3 a,h4 a,h5 a,h6 a{color: hotpink;}h1 a:hove ...
```
These changes need careful reviewing to ensure the expected selectors are accurate and the actual output is also suitable.

adamziel replied:
It looks reasonable to me

getdave commented on packages/edit-site/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -390,10 +390,40 @@ export const toStyles = ( tree, blockSelectors, hasBlockGapSupport ) => {
 		const declarations = getStylesDeclarations( styles );
-		if ( declarations.length === 0 ) {
-			return;
+		if ( declarations?.length ) {
+			ruleset = ruleset + `${ selector }{${ declarations.join( ';' ) };}`;
+		}
```
Previously we aborted here. Now we must continue in order to ensure that pseudo selectors can be processed later on.

ramonjd commented:
> As I envisage it this is what I'm describing above when I say Theme JSON will be open but the UI will be limited. Did I understand you correctly?

Yes! I just wanted to make sure I had understood things correctly 😄 

> If we want to restrict this we're actually going to have to write more code than if we don't want to. I'd be interested in wider opinions on this though.

Good point. I'm just wondering if, by being permissive in the styles we allow inside a pseudo block, we'd cause havoc to theme authors later should we ever change or restrict them. 

Maybe some theme.json elders can chime in here 😄 

> This is for individual block styles right? These are only configurable via the UI correct? In which case we'd want to lock these down to a limited subset of values which the project deems are suitable. 

Yeah, I was thinking of anywhere where we'd set a link:hover color (global, blocks styles in the post editor) via the UI. Restricting at this level sounds like a reasonable compromise 👍

getdave committed:
Update sanitize to support pseudo selectors

getdave commented:
> Good point. I'm just wondering if, by being permissive in the styles we allow inside a pseudo block, we'd cause havoc to theme authors later should we ever change or restrict them.
> 
> Maybe some theme.json elders can chime in here 😄

What we're asking here is do we think we should:

1. Allow all possible style rules inside pseudo selector definitions. _**OR**_
2. Restrict to a subset for now.

The former means you can adjust any CSS property on `:hover` just like you can in CSS. 

The later means we’d lock it down to a subset to styling rules being available.

Be aware that this is in theme.json only so it’s very much a dev only feature. I’m proposing when we introduce interactivity states into the Global Styles UI that we restrict to a subset.

MaggieCabrera commented:
/cc @WordPress/block-themers for testing this too

madhusudhand commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -15,6 +15,16 @@
+	 * Note: this will effect both top level and block level elements.
+	 */
+	const VALID_ELEMENT_PSEUDO_SELECTORS = array(
+		'link' => array( ':hover', ':focus' ),
```
in addition to these, most themes style `:active` state too. Can you add it to allow list.

getdave replied:
Yeh I guess that wouldn't be too much of an issue.

getdave replied:
✅

madhusudhand commented:
If it is not auto generated,`schemas/json/theme.json` needs updates to avoid warnings when `theme.json` referenced it. 

```{ "$schema": "https://schemas.wp.org/trunk/theme.json", .... }```

<img width="353" alt="image" src="https://user-images.githubusercontent.com/1935113/174982672-d53f2913-bcb9-451c-a316-2a43c6634077.png">

adamziel commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -223,6 +419,18 @@ private static function get_block_nodes( $theme_json, $selectors = array() ) {
 					);
+
+					// Handle any psuedo selectors for the element.
+					if ( isset( static::VALID_ELEMENT_PSEUDO_SELECTORS[ $element ] ) ) {
```
That's beyond scope of this PR, but I wonder if `get_style_nodes` and `get_block_nodes` could be unified somehow – the logic is quite similar.

getdave replied:
Many of the functions in this class are become very large and not abiding by single responsibility principle. A refactor could be a good pairing exercise.

adamziel commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -238,11 +446,32 @@ private static function get_block_nodes( $theme_json, $selectors = array() ) {
+
+		// Get a reference to element name from path.
+		// $block_metadata['path'] = array('styles','elements','link');
+		$current_element = $block_metadata['path'][ count( $block_metadata['path'] ) - 1 ];
```
I added an `echo` here like this:

```php
echo '['.implode("][", $block_metadata['path'])."]\n";
```

And these are all the paths it printed:

```
[styles][blocks][core/audio]
[styles][blocks][core/button]
[styles][blocks][core/table]
[styles][blocks][core/video]
[styles][blocks][core/embed]
[styles][blocks][core/image]
[styles][blocks][core/code]
[styles][blocks][core/gallery]
[styles][blocks][core/list]
[styles][blocks][core/navigation]
[styles][blocks][core/post-title]
[styles][blocks][core/post-date]
[styles][blocks][core/pullquote]
[styles][blocks][core/search]
[styles][blocks][core/separator]
[styles][blocks][core/quote]
[styles][blocks][core/site-tagline]
[styles][blocks][core/site-title]
[styles]
[styles][elements][h1]
[styles][elements][h2]
[styles][elements][h3]
[styles][elements][h4]
[styles][elements][h5]
[styles][elements][h6]
[styles][elements][link]
[styles][elements][link]
```

I'm a bit worried about the element-specific logic reasoning about the blocks and other keys as well. This looks safer to me:

```suggestion
		// Make sure that $block_metadata['path'] describes an element node, like ['styles', 'element', 'link'].
		// Skip non-element paths like just ['styles']
		$is_processing_element = count( $block_metadata['path'] ) === 3 && $block_metadata['path'][1] === 'elements';
		$current_element = $is_processing_element ? $block_metadata['path'][ count( $block_metadata['path'] ) - 1 ] : nulll;
```

What do you think?

getdave replied:
Yes that is a good point. This logic is somewhere else in the class as well I think 🤔

getdave replied:
Whoops this broke the unit tests. We're now not getting the block level `elements` because we're checking hard coded path indexes.

getdave replied:
@adamziel I pushed a change which fixes the tests. Does that work for you?

adamziel replied:
Good catch and your fix looks great!

adamziel commented on packages/edit-site/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -390,10 +390,40 @@ export const toStyles = ( tree, blockSelectors, hasBlockGapSupport ) => {
+			( [ key ] ) => key.startsWith( ':' )
+		);
+
+		if ( psuedoSelectorStyles?.length ) {
```
Code style nit:

```suggestion
		if ( ! psuedoSelectorStyles?.length ) {
			return
		}
```

Less nesting, less thinking required.

getdave replied:
I haven't looked in detail but do we definitely want to exit the function here?

adamziel replied:
We're inside of a `forEach` here and this is the last operation so I'd say it's fine to exit. If we want to have multiple segments, we could call `forEach` multiple times like `.forEach(/* do x */).forEach(/* do y */)`. instead of making it one mega function with `if(x) { /* lots of code */ } if(y) { /* lots of code */ }`. But that's beyond the scope of this PR, feel free to make the call on whether you'd like to keep the nesting or not – I'm just expressing opinions, not blockers :-)

getdave replied:
I looked at this and realised we're in the loop (like you said). All good. I agree let's exit early.

getdave commented:
> If it is not auto generated,`schemas/json/theme.json` needs updates to avoid warnings when `theme.json` referenced it.
> 
> `{ "$schema": "https://schemas.wp.org/trunk/theme.json", .... }`
> 

I have no idea how all that works. I wonder who would know?

adamziel commented on packages/edit-site/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -390,10 +390,40 @@ export const toStyles = ( tree, blockSelectors, hasBlockGapSupport ) => {
+
+					ruleset = ruleset + psuedoRule;
+				}
+			} );
```
Code style nit, this should be easier to process:

```js
/**
 * Transforms a selector like
 *    h1 a, h2 a
 * Into one with pseudoclasses like
 *    h1 a:hover, h2 a:hover
 */
function addPseudoClass( selector, pseudoKey ) {
	return (
		selector
			.split( ',' )
			.map( ( sel ) => sel + pseudoKey )
			.join( ',' )
	);
}

for( const [ pseudoClass, pseudoRule ] of psuedoSelectorStyles ) {
	const selectorWithPseudoClass = addPseudoClass( selector, pseudoClass );
	const styleDeclarations = getStylesDeclarations( pseudoRule )?.join(';')
	if ( styleDeclarations ) {
		ruleset += `${ selectorWithPseudoClass }{${ styleDeclarations };}`;
	}
}
```

What I did is:

* Gave a name to the `addPseudoselector` part of the operation
* Replaced the foreach with a `for` loop – to me, it communicates the intention clearer in case of imperative operations.
* Renamed "pseudo key" to "pseudo class" which is how the CSS spec refers to `:hover` and such

getdave replied:
I'd agree with you other than I've just seen https://github.com/WordPress/gutenberg/pull/41822#discussion_r903593083 which is doing much the same.

I feel like we should allow both of these to land then land on an abstraction.

What do you think @scruffian @adamziel ?

adamziel replied:
Sounds good!

adamziel reviewed (approved):
I left a few nits, but it works and looks great in general. Thank you for all your hard work here @getdave !

getdave committed:
Add :active as valid selector

See https://github.com/WordPress/gutenberg/pull/41786#discussion_r903434915

getdave committed:
Check path is an element path

Co-authored-by: Adam Zielinski <adam@adamziel.com>

getdave committed:
Fix null and yoda condition

getdave committed:
Fix check for elements to avoid relying on hardcoded array indexes

Resolves https://github.com/WordPress/gutenberg/pull/41786#discussion_r903602186

getdave committed:
Exit loop early

getdave committed:
Fix linting

mikachan commented:
> I have no idea how all that works. I wonder who would know?

I'm not sure I'm this person, so someone please correct me if I'm wrong... But I believe you need to edit [this file](https://github.com/WordPress/gutenberg/blob/trunk/schemas/json/theme.json) with the new property (`:hover`) information. There's an example of an update like this [here](https://github.com/WordPress/gutenberg/pull/37762).

getdave committed:
Fix offset bug when no pseudo selector present

scruffian committed:
Specifically declare that links should be underlined

scruffian commented:
In 82f5a3760df85b4ad5fccdc8bdaec4c4518dfe07 I added a rule to theme.json for all link elements to be underlined. This is the default style in the browser, but I thought it was important to specifically declare it so that the intention is clear and when  we add the text decoration control to the UI, the correct setting will be selected.

scruffian reviewed (approved):
I think we should merge this and then continue testing it out in more themes.

mburridge commented:
Added the Needs Dev Note label in case this needs a dev note (either individual or as part of a "misc" dev note) for WP 6.1 release.

scruffian commented:
I mentioned it in this post: https://make.wordpress.org/core/?p=98557&preview=true

mburridge commented:
Thanks @scruffian. Might it be worth adding the #dev-notes tag to that post?

getdave commented:
> https://make.wordpress.org/core/?p=98557&preview=true

I can't see that post @scruffian.

getdave commented:
# WP 6.1 Dev note

In 6.1 support has been added to allow developers to define styles for pseudo classes (e.g. `:hover`, `:focus`) for a limited set of elements via Theme JSON.

Support for this feature is available both for root level elements and also for elements on per-block basis. 

To add these styles, include the appropriate pseudo class (prepended with a semicolon) in your style definitions and then nest additional styles rules _within_ this to define how your element should look when the pseudo class is active:

```json
{
    "elements": {
        "link": {
            "color": {
                "text": "green"
            },
            ":hover": {
                "color": {
                    "text": "hotpink"
                }
            }
        }
    }
}
```

This will result in the following CSS being output in a Theme (please note actual output may vary):

```css
a {
    color: green;
}

a:hover {
    color: hotpink;
}
```

This functions similarly at a block level, the only difference being the rules should be applied to the `elements` nested within the block you wish to target:

```json
{
    "blocks": {
        "core/group": {
             "elements": {
                "link": {
                    ":hover": {
                        "color": {
                            "text": "red"
                        }
                    }
                }
            }
        }
    }
}
```

In this case `<a>` elements within a `core/group` block will have `red` text on `:hover` - this results in the following CSS (actual output may vary):

```css
.wp-block-group a:hover {
    color: red;
}
```

At the current time support for this feature is intentionally limited to only `link` (i.e. `<a>`) and `button` elements to avoid a11y and usability concerns. It may be extended in future.

youknowriad commented on lib/compat/wordpress-6.1/theme.json:
```diff
@@ -250,6 +250,11 @@
+			"link": {
+				"typography": {
+					"textDecoration": "underline"
+				}
```
Do you recall why we have to explicitly set the link style to "underline" in the core theme.json file? I mean the default behavior of browser is to underline links, so why we had to add this explicitly?

getdave replied:
I really can't I'm afraid. If I had to guess, I'd say that for a11y reasons Navigation links would need to be underlined and perhaps default styling at the time was not not have them underlined? It may be that things have changed and browser defaults are now allowed.

Perhaps @MaggieCabrera or @mikachan would know?

MaggieCabrera replied:
I can't think of a reason right now! Maybe we had some reset CSS back then of some kind? Nowadays it seems like we could go without it
